File: frontend/src/api/community.ts
import axios from 'axios';

const API_BASE_URL = __DEV__ 
  ? 'http://10.56.56.21:8000'
  : 'https://your-production-url.com';

export interface CommunityPost {
  id: string;
  title: string;
  description: string;
  file_type: 'image' | 'video' | 'dataset' | 'file' | 'text';
  file_url?: string;
  file_name?: string;
  user_id: string;
  created_at: string;
  updated_at: string;
}

export interface UploadCommunityFileRequest {
  file: {
    uri: string;
    name: string;
    type: string;
  };
  file_type: 'image' | 'video' | 'dataset';
  user_id: string;
  title: string;
  description: string;
}

/**
 * 커뮤니티에 파일 업로드
 * React Native의 axios가 여러 텍스트 필드가 포함된 FormData를 처리하지 못하는 문제 해결
 * 해결책: analyzeVideo와 유사하게 최소한의 필드만 사용
 * - user_id (필수)
 * - metadata (JSON 문자열로 모든 메타데이터 포함)
 * - file (파일이 있는 경우만)
 */
export async function uploadCommunityFile(formData: FormData): Promise<CommunityPost> {
  try {
    console.log('[uploadCommunityFile] ====== 파일 업로드 요청 시작 ======');
    const apiUrl = `${API_BASE_URL}/community/upload`;
    console.log('[uploadCommunityFile] API URL:', apiUrl);
    console.log('[uploadCommunityFile] FormData 필드: user_id + file (analyzeVideo와 완전히 동일한 방식)');
    
    // analyzeVideo와 완전히 동일한 방식: fetch API를 먼저 사용
    try {
      // 먼저 fetch API로 시도 (React Native에서 더 안정적일 수 있음)
      console.log('[uploadCommunityFile] fetch API로 요청 전송 시도...');
      const fetchResponse = await fetch(apiUrl, {
        method: 'POST',
        body: formData,
        headers: {
          'Accept': 'application/json',
          // Content-Type은 설정하지 않음 - fetch가 자동으로 boundary 설정
        },
      });

      if (!fetchResponse.ok) {
        const errorText = await fetchResponse.text();
        console.error('[uploadCommunityFile] fetch 응답 실패:', fetchResponse.status, errorText);
        throw new Error(`서버 오류: ${fetchResponse.status} - ${errorText}`);
      }

      const responseData = await fetchResponse.json();
      console.log('[uploadCommunityFile] 응답 받음 (fetch):', fetchResponse.status);
      console.log('[uploadCommunityFile] 응답 데이터:', JSON.stringify(responseData).substring(0, 500));
      console.log('[uploadCommunityFile] 업로드 성공');
      return responseData;
    } catch (fetchError: any) {
      console.warn('[uploadCommunityFile] fetch API 실패, axios로 재시도:', fetchError?.message);
      
      // fetch 실패 시 axios로 재시도
      const response = await axios.post(
        apiUrl,
        formData,
        {
          headers: { 
            // Content-Type을 명시하지 않음 - axios가 자동으로 multipart/form-data + boundary 설정
            'Accept': 'application/json',
          },
          timeout: 300000, // 5분 타임아웃
          maxContentLength: Number.MAX_SAFE_INTEGER,
          maxBodyLength: Number.MAX_SAFE_INTEGER,
        }
      );
      
      console.log('[uploadCommunityFile] 응답 받음 (axios):', response.status);
      console.log('[uploadCommunityFile] 응답 데이터:', JSON.stringify(response.data).substring(0, 500));
      console.log('[uploadCommunityFile] 업로드 성공');
      return response.data;
    }
  } catch (error: any) {
    console.error('[uploadCommunityFile] ====== 파일 업로드 실패 ======');
    console.error('커뮤니티 파일 업로드 실패:', error);
    console.error('오류 상세:', {
      message: error?.message,
      code: error?.code,
      name: error?.name,
      stack: error?.stack?.substring(0, 500),
      response: error?.response?.data,
      status: error?.response?.status,
      request: error?.request ? 'request exists' : 'no request',
    });
    
    // Network Error는 요청이 전송되지 않았음을 의미
    if (error.code === 'ECONNREFUSED' || error.message === 'Network Error' || error.code === 'ERR_NETWORK' || error.message === 'Network request failed') {
      console.error('[uploadCommunityFile] ⚠️ 요청이 서버에 도달하지 않음');
      console.error('[uploadCommunityFile] 백엔드 서버 연결 확인 필요');
      console.error('[uploadCommunityFile] API URL:', `${API_BASE_URL}/community/upload`);
      throw new Error('백엔드 서버에 연결할 수 없습니다. 서버가 실행 중인지 확인하세요.');
    }
    
    if (error.code === 'ECONNABORTED' || error.message?.includes('timeout')) {
      throw new Error('파일 업로드 시간이 초과되었습니다. 파일 크기를 확인하세요.');
    }
    
    throw error;
  }
}

/**
 * 커뮤니티 게시글 목록 조회
 */
export async function getCommunityPosts(limit: number = 50, offset: number = 0): Promise<CommunityPost[]> {
  try {
    const response = await axios.get(
      `${API_BASE_URL}/community/posts`,
      {
        params: {
          limit,
          offset,
        },
      }
    );
    return response.data;
  } catch (error) {
    console.error('커뮤니티 게시글 조회 실패:', error);
    throw error;
  }
}

/**
 * 특정 게시글 조회
 */
export async function getCommunityPost(postId: string): Promise<CommunityPost> {
  try {
    const response = await axios.get(`${API_BASE_URL}/community/posts/${postId}`);
    return response.data;
  } catch (error) {
    console.error('게시글 조회 실패:', error);
    throw error;
  }
}

/**
 * 게시글 삭제
 */
export async function deleteCommunityPost(postId: string): Promise<void> {
  try {
    await axios.delete(`${API_BASE_URL}/community/posts/${postId}`);
  } catch (error) {
    console.error('게시글 삭제 실패:', error);
    throw error;
  }
}

File: backend/app/api/endpoints/community.py
from fastapi import APIRouter, UploadFile, File, Form, HTTPException, Query
from typing import List, Optional
from datetime import datetime
import uuid
import os
import asyncio
from concurrent.futures import ThreadPoolExecutor
from app.core.firebase import db, bucket
import firebase_admin
from firebase_admin import firestore
from firebase_admin import storage

# ThreadPoolExecutor for Firestore operations
_executor = ThreadPoolExecutor(max_workers=5)

router = APIRouter()

# 커뮤니티 게시글 컬렉션 이름
COMMUNITY_COLLECTION = "community_posts"


@router.post("/upload", summary="커뮤니티 파일 업로드")
async def upload_community_file(
    file: Optional[UploadFile] = File(None),
    file_type: Optional[str] = Form("text"),
    user_id: str = Form(...),
    title: Optional[str] = Form(""),
    description: Optional[str] = Form(""),
    metadata: Optional[str] = Form(None),  # JSON 문자열로 메타데이터 전송 (선택적)
):
    """
    커뮤니티에 파일 업로드 (사진, 영상, 데이터셋, 텍스트만)
    한글 제목 및 본문 지원
    metadata 필드가 있으면 JSON으로 파싱하여 title, description, file_type 사용
    """
    try:
        # metadata가 있으면 JSON 파싱하여 사용 (React Native FormData 처리 문제 해결)
        if metadata:
            try:
                import json
                metadata_dict = json.loads(metadata)
                title = metadata_dict.get("title", title or "")
                description = metadata_dict.get("description", description or "")
                file_type = metadata_dict.get("file_type", file_type or "text")
                print(f"[community/upload] metadata에서 파싱: title={title}, description 길이={len(description)}, file_type={file_type}")
            except Exception as e:
                print(f"[community/upload] metadata 파싱 실패, 기본값 사용: {e}")
        
        # 파일명에서 메타데이터 파싱 (파일명 형식: metadata|originalname)
        if file and file.filename:
            try:
                import json
                import urllib.parse
                filename_parts = file.filename.split('|', 1)
                if len(filename_parts) == 2:
                    # 파일명에 메타데이터가 인코딩되어 있음
                    encoded_metadata = filename_parts[0]
                    original_filename = filename_parts[1]
                    decoded_metadata = urllib.parse.unquote(encoded_metadata)
                    metadata_dict = json.loads(decoded_metadata)
                    title = metadata_dict.get("title", title or "")
                    description = metadata_dict.get("description", description or "")
                    file_type = metadata_dict.get("file_type", file_type or "text")
                    # 원본 파일명 복원
                    file.filename = original_filename
                    print(f"[community/upload] 파일명에서 메타데이터 파싱: title={title}, description 길이={len(description)}, file_type={file_type}, original_filename={original_filename}")
            except Exception as e:
                print(f"[community/upload] 파일명에서 메타데이터 파싱 실패 (정상일 수 있음): {e}")
        
        print(f"[community/upload] ====== 파일 업로드 요청 받음 ======")
        print(f"[community/upload] 제목: {title}, 본문 길이: {len(description)}")
        print(f"[community/upload] user_id: {user_id}, file_type: {file_type}")
        print(f"[community/upload] file is None: {file is None}")
        # 파일 타입 검증 (모든 파일 타입 허용)
        valid_file_types = ["image", "video", "dataset", "file", "text"]
        if file_type not in valid_file_types:
            file_type = "file"  # 기본값으로 설정
        
        file_url = None
        file_name = None
        file_content = None
        file_size = 0
        
        # 파일이 있는 경우에만 업로드
        if file is not None:
            try:
                # 파일 정보 확인
                print(f"파일 업로드 시작 - filename: {file.filename}, content_type: {file.content_type}")
                
                if hasattr(file, 'filename') and file.filename:
                    # 파일 읽기 (청크 단위로 읽어서 메모리 효율성 향상)
                    file_content = await file.read()
                    file_size = len(file_content) if file_content else 0
                    
                    print(f"파일 크기: {file_size} bytes ({file_size / 1024 / 1024:.2f} MB)")
                    
                    # 파일 크기 제한 체크 (500MB)
                    max_file_size = 500 * 1024 * 1024  # 500MB
                    if file_size > max_file_size:
                        raise HTTPException(status_code=413, detail=f"파일 크기가 너무 큽니다. 최대 {max_file_size / 1024 / 1024}MB까지 허용됩니다.")
                    
                    # 파일명 생성 (UUID 사용)
                    file_extension = os.path.splitext(file.filename)[1] if file.filename else ""
                    unique_filename = f"{uuid.uuid4()}{file_extension}"
                    file_name = file.filename
                    
                    # Firebase Storage 경로 설정
                    storage_path = f"community/{file_type}/{unique_filename}"
                    
                    print(f"Firebase Storage 업로드 시작: {storage_path}")
                    
                    # Firebase Storage에 업로드
                    blob = bucket.blob(storage_path)
                    blob.upload_from_string(file_content, content_type=file.content_type or 'application/octet-stream')
                    blob.make_public()
                    file_url = blob.public_url
                    
                    print(f"Firebase Storage 업로드 완료: {file_url}")
            except HTTPException:
                raise
            except Exception as file_error:
                print(f"파일 업로드 오류: {file_error}")
                import traceback
                traceback.print_exc()
                # 파일 업로드 실패해도 텍스트 글은 작성 가능
                file_url = None
                file_name = None
        
        # 게시글 데이터 생성
        post_id = str(uuid.uuid4())
        post_data = {
            "id": post_id,
            "title": title,
            "description": description,
            "file_type": file_type,
            "file_url": file_url,
            "file_name": file_name,
            "file_size": file_size,
            "user_id": user_id,
            "created_at": datetime.now().isoformat(),
            "updated_at": datetime.now().isoformat(),
        }
        
        # Firestore에 저장 (타임아웃 처리)
        try:
            def save_to_firestore():
                db.collection(COMMUNITY_COLLECTION).document(post_id).set(post_data)
            
            # 10초 타임아웃으로 실행
            loop = asyncio.get_event_loop()
            await asyncio.wait_for(
                loop.run_in_executor(_executor, save_to_firestore),
                timeout=10.0
            )
        except asyncio.TimeoutError:
            print(f"Firestore 저장 타임아웃 (10초 초과)")
            raise HTTPException(status_code=503, detail="데이터베이스 연결 시간이 초과되었습니다. 잠시 후 다시 시도해주세요.")
        except Exception as firestore_error:
            print(f"Firestore 저장 오류: {firestore_error}")
            import traceback
            traceback.print_exc()
            raise HTTPException(status_code=500, detail=f"데이터 저장 실패: {str(firestore_error)}")
        
        return post_data
        
    except HTTPException:
        raise
    except Exception as e:
        print(f"커뮤니티 파일 업로드 오류: {e}")
        raise HTTPException(status_code=500, detail=f"파일 업로드 실패: {str(e)}")


@router.get("/posts", summary="커뮤니티 게시글 목록 조회")
async def get_community_posts(
    limit: int = Query(50, ge=1, le=100, description="조회할 게시글 수"),
    offset: int = Query(0, ge=0, description="건너뛸 게시글 수"),
    file_type: Optional[str] = Query(None, description="파일 타입 필터 (image, video, dataset)"),
):
    """
    커뮤니티 게시글 목록 조회
    """
    try:
        # Firestore 쿼리 생성 (타임아웃 처리)
        query = db.collection(COMMUNITY_COLLECTION)
        
        # 파일 타입 필터 적용 (모든 파일 타입 지원)
        if file_type:
            query = query.where("file_type", "==", file_type)
        
        # 최신순 정렬 및 페이지네이션 (타임아웃 처리)
        posts = []
        try:
            # 타임아웃 설정 (10초)
            def fetch_posts():
                docs = query.order_by("created_at", direction=firestore.Query.DESCENDING).limit(limit + offset).stream()
                result = []
                for i, doc in enumerate(docs):
                    if i >= offset:
                        post_data = doc.to_dict()
                        result.append(post_data)
                return result
            
            # 10초 타임아웃으로 실행
            loop = asyncio.get_event_loop()
            posts = await asyncio.wait_for(
                loop.run_in_executor(_executor, fetch_posts),
                timeout=10.0
            )
        except asyncio.TimeoutError:
            print(f"Firestore 쿼리 타임아웃 (10초 초과)")
            # 타임아웃 시 빈 리스트 반환
            return []
        except Exception as query_error:
            print(f"Firestore 쿼리 오류: {query_error}")
            import traceback
            traceback.print_exc()
            # Firestore 연결 실패 시 빈 리스트 반환
            return []
        
        return posts
        
    except HTTPException:
        raise
    except Exception as e:
        print(f"커뮤니티 게시글 조회 오류: {e}")
        import traceback
        traceback.print_exc()
        # 오류 발생 시 빈 리스트 반환 (서버 오류로 앱이 멈추지 않도록)
        return []


@router.get("/posts/{post_id}", summary="특정 게시글 조회")
async def get_community_post(post_id: str):
    """
    특정 게시글 조회
    """
    try:
        # Firestore에서 게시글 조회 (타임아웃 처리)
        try:
            def fetch_post():
                return db.collection(COMMUNITY_COLLECTION).document(post_id).get()
            
            # 10초 타임아웃으로 실행
            loop = asyncio.get_event_loop()
            doc = await asyncio.wait_for(
                loop.run_in_executor(_executor, fetch_post),
                timeout=10.0
            )
        except asyncio.TimeoutError:
            print(f"Firestore 조회 타임아웃 (10초 초과)")
            raise HTTPException(status_code=503, detail="데이터베이스 연결 시간이 초과되었습니다. 잠시 후 다시 시도해주세요.")
        except Exception as query_error:
            print(f"Firestore 조회 오류: {query_error}")
            import traceback
            traceback.print_exc()
            raise HTTPException(status_code=503, detail=f"데이터베이스 연결 실패: {str(query_error)}")
        
        if not doc.exists:
            raise HTTPException(status_code=404, detail="게시글을 찾을 수 없습니다.")
        
        return doc.to_dict()
        
    except HTTPException:
        raise
    except Exception as e:
        print(f"게시글 조회 오류: {e}")
        import traceback
        traceback.print_exc()
        raise HTTPException(status_code=500, detail=f"게시글 조회 실패: {str(e)}")


@router.delete("/posts/{post_id}", summary="게시글 삭제")
async def delete_community_post(post_id: str, user_id: str = Query(..., description="사용자 ID")):
    """
    게시글 삭제 (작성자만 삭제 가능)
    """
    try:
        doc_ref = db.collection(COMMUNITY_COLLECTION).document(post_id)
        doc = doc_ref.get()
        
        if not doc.exists:
            raise HTTPException(status_code=404, detail="게시글을 찾을 수 없습니다.")
        
        post_data = doc.to_dict()
        
        # 작성자 확인
        if post_data.get("user_id") != user_id:
            raise HTTPException(status_code=403, detail="게시글을 삭제할 권한이 없습니다.")
        
        # Firebase Storage에서 파일 삭제
        file_url = post_data.get("file_url")
        if file_url:
            try:
                # URL에서 파일 경로 추출
                # 예: https://storage.googleapis.com/bucket/path/to/file
                # 또는 https://firebasestorage.googleapis.com/v0/b/bucket/o/path%2Fto%2Ffile
                if "firebasestorage.googleapis.com" in file_url:
                    # Firebase Storage URL 파싱
                    import urllib.parse
                    path_part = file_url.split("/o/")[-1].split("?")[0]
                    file_path = urllib.parse.unquote(path_part)
                    blob = bucket.blob(file_path)
                    blob.delete()
            except Exception as e:
                print(f"파일 삭제 오류 (무시): {e}")
        
        # Firestore에서 게시글 삭제
        doc_ref.delete()
        
        return {"message": "게시글이 삭제되었습니다."}
        
    except HTTPException:
        raise
    except Exception as e:
        print(f"게시글 삭제 오류: {e}")
        raise HTTPException(status_code=500, detail=f"게시글 삭제 실패: {str(e)}")

